src/inmanta/data/__init__.py

```diff
@@ -2630,21 +2656,33 @@ async def get_list_without_details(
 
     @classmethod
     async def get_by_id(
-        cls: Type[TBaseDocument],
+        cls,
         doc_id: uuid.UUID,
         connection: Optional[asyncpg.connection.Connection] = None,
         details: bool = True,
-    ) -> Optional[TBaseDocument]:
+    ) -> Optional["Environment"]:
         """
         Get a specific environment based on its ID
 
         :return: An instance of this class with its fields filled from the database.
         """
+        if doc_id in cls.__instance_cache and details in cls.__instance_cache[doc_id]:
```

Comment:
```suggestion
        if connection is None and doc_id in cls.__instance_cache and details in cls.__instance_cache[doc_id]:
```
I'm not sure about this but I wonder if we could get strange inconsistency errors if we allow the cache to live accross transactions.